mining.go

```diff
@@ -279,6 +280,107 @@ func containsTxIns(txs []*dcrutil.Tx, tx *dcrutil.Tx) bool {
 	return false
 }
 
+// blockWithNumVotes is a block with the number of votes currently present
+// for that block. Just used for sorting.
+type blockWithNumVotes struct {
+	Hash     chainhash.Hash
+	NumVotes uint16
+}
+
+// byNumberOfVotes implements sort.Interface to sort a slice of blocks by their
+// number of votes.
+type byNumberOfVotes []*blockWithNumVotes
+
+// Len returns the number of elements in the slice.  It is part of the
+// sort.Interface implementation.
+func (b byNumberOfVotes) Len() int {
+	return len(b)
+}
+
+// Swap swaps the elements at the passed indices.  It is part of the
+// sort.Interface implementation.
+func (b byNumberOfVotes) Swap(i, j int) {
+	b[i], b[j] = b[j], b[i]
+}
+
+// Less returns whether the block with index i should sort before the block with
+// index j.  It is part of the sort.Interface implementation.
+func (b byNumberOfVotes) Less(i, j int) bool {
+	return b[i].NumVotes < b[j].NumVotes
+}
+
+// SortParentsByVotes takes a list of block header hashes and sorts them
+// by the number of votes currently available for them in the votes map of
+// mempool.  It then returns all blocks that are eligible to be used (have
+// at least a majority number of votes) sorted by number of votes, descending.
+//
+// This function is safe for concurrent access.
+func SortParentsByVotes(mp *txMemPool, currentTopBlock chainhash.Hash, blocks []chainhash.Hash, params *chaincfg.Params) []chainhash.Hash {
+	// Return now when no blocks were provided.
+	lenBlocks := len(blocks)
+	if lenBlocks == 0 {
+		return nil
+	}
+
+	// Fetch the vote metadata for the provided block hashes from the
+	// mempool and filter out any blocks that do not have the minimum
+	// required number of votes.
+	minVotesRequired := (params.TicketsPerBlock / 2) + 1
+	voteMetadata := mp.VotesForBlocks(blocks)
+	filtered := make([]*blockWithNumVotes, 0, lenBlocks)
+	for i := range blocks {
+		numVotes := uint16(len(voteMetadata[i]))
+		if numVotes >= minVotesRequired {
+			filtered = append(filtered, &blockWithNumVotes{
+				blocks[i], numVotes})
```

Comment:
I'm not a fan in general, but the code was already that way, so I was attempting to change as little as possible to decouple it and correct the things that didn't make sense.

I will happily change it.